### PR TITLE
feat: `partition`/`partitionRecord` utils

### PR DESCRIPTION
**Pull request by agurtovoy**
Pure addition, prep for https://github.com/makeswift/makeswift/pull/814.

### CODE REVIEW

File: packages/runtime/src/utils/__tests__/partition.test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  return typeof value === 'number'
+}
+
+describe('partition', () => {
+  it('splits the array into two based on a predicate', () => {
+    const array = [1, 2, 3, 4, 5]
+    const [even, odd] = partition(array, x => x % 2 === 0)
+    expect(even).toStrictEqual([2, 4])
+    expect(odd).toStrictEqual([1, 3, 5])
+  })
```
Comment (arvinpoddar) [inline packages/runtime/src/utils/__tests__/partition.test.ts:15]: I'd consider adding a test to make sure the method returns an empty partition when none of the values match the predicate. I know the function signature protects us from someone accidentally modifying it to return `null` or `undefined`, but I feel that it's a helpful addition.

File: packages/runtime/src/utils/partition.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+type Predicate<T, U extends T> = (value: T) => value is U
+
+export function partition<T, U extends T>(
+  array: readonly T[],
+  predicate: Predicate<T, U>,
+): [U[], Exclude<T, U>[]]
+
+export function partition<T>(array: readonly T[], predicate: (value: T) => boolean): [T[], T[]]
+
+export function partition(
```
Comment (arvinpoddar) [inline packages/runtime/src/utils/partition.ts:10]: May be helpful to add a JSDoc string clarifying that the first returned array is the matches, and that the results are returned in stable order.

**Review by arvinpoddar — APPROVED** [review]

File: packages/runtime/src/utils/__tests__/partition.test.ts (hunk 1 above)
Comment (agurtovoy) [reply packages/runtime/src/utils/__tests__/partition.test.ts:15]: Good idea.

File: packages/runtime/src/utils/partition.ts (hunk 2 above)
Comment (agurtovoy) [reply packages/runtime/src/utils/partition.ts:10]: Good point, we should probably document all of the general-purpose "utils" functions.

Comment (agurtovoy) [reply packages/runtime/src/utils/partition.ts:10]: This particular convention for `partition` is a de facto standard, but docs never hurt.

**Review by migueloller — DISMISSED** [review]
LGTM 👍🏻

**Comment by migueloller** [comment]
Minor note, it was a bit weird to have to ignore changes from another PR. I really dislike how GitHub doesn't allow to do a review commit by commit and instead forces one to review all files at once 🙄

**Comment by agurtovoy** [comment]
@migueloller My fault, I haven't rebased on `main` after merging the other one. Rebased now.
